### ISSUE TITLE
Handle Goal Creation Latency

### ISSUE THREAD

**Issue by dvehar** [devcolor/majorkey#27]
When I clicked create it took a good 3 or 4 seconds to see the success notification. Meanwhile I clicked the create button again causing a duplicate. This is not good...

TODO:
- [ ] Make "Create" button disabled while submission is being processed

Maybe we should:
- [ ] Make the goal creation processing faster

**Comment by makinde** [comment]
Yep, button should be disabled.
Sadly, the delay in seeing the dialog is the time it takes for /directory to load. The endpoint for goal creation is doing a redirect back to the directory with that banner on top. Need to optimize the sloppy queries there.
Aside, do you think the goals should live on their own page?






On Tue, Apr 12, 2016 at 3:11 PM -0700, "Desmond Vehar" <notifications@github.com> wrote:












When I clicked create it took a good 3 or 4 seconds to see the success notification. Meanwhile I clicked the create button again.

TODO:



 Make "Create" button disabled while submission is being processed




Maybe we should:



 Make the goal creation processing faster

**Comment by dvehar** [comment]
@makinde  I like that the goals are right there on the top. It makes it the primary focus of the /directory page. If that is not the intention then maybe we should just have a link to "Your Commitments", or show a few upcoming goals based on date and then have a link to see the rest of the goals.

**Comment by makinde** [comment]
@dvehar Yep, that the exact intent. 

I should do some profiling, but I'm guessing the problem code is in this query...
`      return models.Goal
        .find({ owner : { $in: member_ids }})
        .or([
          { status: 'OPEN' },
          { status: 'COMPLETED', updated_at: {$gte: moment().subtract(30, 'd').toDate() }},
          { status: 'ABANDONED', updated_at: {$gte: moment().subtract(20, 'm').toDate() }}
        ])
        .exec();`

I think I can simplify it down quite a bit. The /directory page has an avg load time of 250ms but routinely degrades to ~1-1.5s. 

@dvehar, you want to tack a crack at disabling the create button on click?

**Comment by dvehar** [comment]
Yea I'll take a stab at it.

**Comment by makinde** [comment]
@dvehar I looked factored the above query to just make it simpler (win), but the page is still slow. I don't have the bandwidth to actually profile this right now, but I'd guess that we can get away caching all of our data in process to speed things up quite a bit. We're dealing with basically no data here. It'll be under 1MB or a while. If you have the time, feel free to look into solutions.

**Comment by astonm** [comment]
I'm seeing ~400ms to create the new goal plus ~800ms to reload the page for a total of 1s+ latency.

I think the better solve here is, when you create a goal, don't reload the entire directory page--just add a row client-side. We can even do it optimistically (that is, before we hear back from the server that it succeeded) then just handle a server failure by undoing that newly added row and showing an error. But even without the optimistic rendering, we can make this 3x faster.

**Comment by makinde** [comment]
Yeah, those response times are grossly slower than they should be. Something is up.

For simplicity, it'd be great to see if we can keep all of our rendering in the same place. We're somewhere like where facebook way in 2008. Adding bits of client-side rendering to keep in sync with server side rendering led us to a bad place.

That goal creation endpoint is doing a single DB write and returning (though the login takes another read). If we got that endpoint under 40ms and the directory load under 200ms, would that be fast enough to keep it the way it is? Or do you want to move to react and ditch nunjucks all together?

**Comment by astonm** [comment]
I'm seeing around 75ms just to run `models.Goal(new_goal).save()`, so getting the endpoint to 40ms will take optimizing something in Mongo / mongoose.

**Comment by makinde** [comment]
Yeah, just mind blowing to me that latency is that high. I checked, our mongoDB is in the same AWS region as our heroku stuff. Tried to dig into a bunch of our queries is made sure they were all using indexing. The profiling info I'm seeing from the mongo side is saying everything is running really fast. It doesn't have these queries for long. It'd be crazy/sad if mongoose/all the middleware is churning up all that time.

We have so little data we can afford to cache everything in process if need be. There are a few libs to patch mongoose.

We can also ask in the slack channel if anyone's actually used Mongbo and knows what to expect/how to tune.

**Comment by dvehar** [comment]
I'll get started soon

**Comment by makinde** [comment]
Okay, I setup NewRelic last night cause I couldn't sleep :P I've got two good traces that tell me a bit more, but also nothing at all.

Below are two slow requests from the screenshots of request profiles for the directory. You see the queries go off, 2 of the 3 return quickly, the third one stalls for 100+ms. The rub: it's a different query each time. I ran `explain()` on each of the queries and they are hitting indices and all this data should easily be in memory. So sometimes a random query will stall out. This is silliness. I'm gonna try copying the db to composedb, another provider.

![screen shot 2016-05-18 at 11 54 44 am](https://cloud.githubusercontent.com/assets/157415/15371666/7a45a1e4-1cf1-11e6-88d7-3207bc7024a9.png)
![screen shot 2016-05-18 at 12 07 29 pm](https://cloud.githubusercontent.com/assets/157415/15371674/81a7ad1a-1cf1-11e6-824d-c5dfba32b9ea.png)
